tools/profiling/dbpreader.c

```diff
@@ -382,6 +382,12 @@ static const dbp_event_t *dbp_iterator_next_buffer(dbp_event_iterator_t *it)
     assert( it->current_events_buffer->buffer_type == PROFILING_BUFFER_TYPE_EVENTS );
 
     next_off = it->current_events_buffer->next_buffer_file_offset;
+    if(next_off == -1) {
+        it->current_event.native = NULL;
+        it->current_events_buffer = NULL;
+        return NULL;
+    }
+
     return dbp_iterator_set_offset(it, next_off);
```

Comment:
I'm not sure I see why this is needed. `dbp_iterator_set_offset` releases the current event buffer and sets `it->current_events_buffer` and `it->current_event.native` to `NULL`, then attempts to go to buffer with said offset via `refer_events_buffer`. When `offset == -1` then I think that both the `mmap` and `lseek` methods should fail and return `NULL`. Finally, `dbp_iterator_move_to_event` is called; here, if `it->current_events_buffer` is `NULL` (which it should be since `refer_events_buffer` failed) then `it->current_event.native` is set to `NULL` too. The return value is that of `dbp_iterator_current`, which is `NULL` when either of `it->current_events_buffer` or `it->current_event.native` is `NULL`.

The only way I see a codepath causing issues is if `refer_events_buffer` somehow _doesn't_ fail for `offset == -1`, in which case that would be the correct place to add such a check.

This code also doesn't release the current buffer, so you leak that memory.